### DISCUSSION TITLE
Multi-target tracking from CSV file

### DISCUSSION THREAD

**Discussion by bkev13** [dstl/Stone-Soup#1073] [Q&A] [answered]
I have a CSV file that contains the XYZ and timestamps of a list of objects over a certain period. In this list some objects are there from the beginning till the end, and some join or leave before the beginning and the end. I followed the tutorials and it is unclear how multiple objects will be detected at a specific time, and how to do so. I would appreciate any help you can give me. I want to perform JPDA and use Tutorials 6, 8, 9, 10 and the UAV tracking as references.

And for the class [CSVGroundTruthReader], what is the role of the path_id_field what is the role of path_id_field? In the provided csv files in the tutorials they are all the same number.

**Comment by apiszcz** [comment]
FYSA

https://github.com/dstl/Stone-Soup/issues/971

**Reply by bkev13** [reply]
@apiszcz, thanks for the suggestion, I would look into it.

**Comment by sdhiscocks** [comment]
@bkev13: So just to clarify, does your input file represent a number of true ground truthed objects that you wish to simulate sensor observations for, and then track that? Or is your CSV file observations of the true objects, and you wish to track?

**Reply by bkev13** [reply]
@sdhiscocks: Thanks for you response, my CSV file contains the observations of the true object that I wish to track.

**Reply by sdhiscocks** [reply]
I've not test the below code, but this is roughly what you'd want to look at doing. There are a number of parameters that I've tried to highlight with `?` which will be dependent on your targets and measurements.

```python

target_velocity_noise = ?  # Assumes same all 3-dimensions
measurement_noise = ?  # Assumes same all 3-dimensions

clutter_spatial_density = ?
prob_detect = ?

velocity_prior_covar_value = ?  # Assumes same all 3-dimensions

min_detections = ?  # 3

covariance_limit_for_delete = ?

from stonesoup.reader.generic import CSVDetectionReader
detector = CSVDetectionReader(
    "filename?",
    state_vector_fields=("X?", "Y?", "Z?"),
    time_field="time_field?")

measurement_model_covariance = np.diag([measurement_noise, measurement_noise, measurement_noise])
measurement_model = LinearGaussian(6, [0, 2, 4], measurement_model_covariance)

from stonesoup.models.transition.linear import (
    CombinedLinearGaussianTransitionModel, ConstantVelocity)
transition_model = CombinedLinearGaussianTransitionModel(
    [ConstantVelocity(target_velocity_noise),
     ConstantVelocity(target_velocity_noise),
     ConstantVelocity(target_velocity_noise)])

from stonesoup.predictor.kalman import KalmanPredictor
predictor = KalmanPredictor(transition_model)

from stonesoup.updater.kalman import KalmanUpdater
updater = KalmanUpdater(measurement_model)

from stonesoup.hypothesiser.probability import PDAHypothesiser
hypothesiser = PDAHypothesiser(predictor=predictor,
                               updater=updater,
                               clutter_spatial_density=clutter_spatial_density,
                               prob_detect=prob_detect)

from stonesoup.dataassociator.probability import JPDA
data_associator = JPDA(hypothesiser=hypothesiser)

from stonesoup.dataassociator.neighbour import GNNWith2DAssignment
init_data_associator = GNNWith2DAssignment(hypothesiser)

from stonesoup.deleter.error import CovarianceBasedDeleter
deleter = CovarianceBasedDeleter(covar_trace_thresh=covariance_limit_for_delete)

s_prior_state = GaussianState(
    [[0], [0], [0], [0], [0], [0]],
    np.diag([0, velocity_prior_covar_value,
             0, velocity_prior_covar_value,
             0, velocity_prior_covar_value]))

from stonesoup.initiator.simple import MultiMeasurementInitiator
initiator = MultiMeasurementInitiator(
    prior_state=s_prior_state,
    measurement_model=measurement_model,
    deleter=deleter,
    data_associator=init_data_associator,
    updater=updater,
    min_points=min_detections
)

from stonesoup.tracker.simple import MultiTargetMixtureTracker

tracker = MultiTargetMixtureTracker(
    initiator=initiator,
    deleter=deleter,
    detector=detector,
    data_associator=data_associator,
    updater=updater,
)

detections = set()
tracks = set()
for time, ctracks in tracker:
    detections.update(detection_sim.detections)
    tracks.update(ctracks)
```

**Reply by bkev13** [reply]
Thanks for the reply, this helped me get a start. I just have one more question, what is detection_sim/what is its role? From the tutorials, it seems to me that it is used to simulate detections based on the ground truth. Maybe I should have specified it sooner, but I don't have the ground truth, just the measurements from the sensor in my CSV file.

**Reply by sdhiscocks** [reply]
Made an edit to the above with adding `init_data_associator`, due to multi measurement initiator not supporting JPDA.

So detection sim is simulating detections. In code above it is using CSV detections directly. I had left `detection_sim` in the example above before I edited so may have appeared in email notification if that's where you saw it.

**Reply by bkev13** [reply]
Thank you for the help. I implemented your changes but the tracks set is empty after all the computations. if it helps, here is my script as a Txt file and the associated CSV file.
[soup_solution_v1.txt](https://github.com/user-attachments/files/16560598/soup_solution_v1.txt)
[measurements_snippet1.csv](https://github.com/user-attachments/files/16560433/measurements_snippet1.csv)

**Reply by sdhiscocks** [reply, marked as answer]
[I've had a little play with parameters as an initial start](https://gist.github.com/sdhiscocks/8bf04036ece3bf03bdbad50c52436043). Some truth data to validate and optimise parameters would be useful, but understand that's not always available. I've also used [pyehm plugin](https://pyehm.readthedocs.io/en/latest/index.html) to speed up JPDA, but note the patent required for commercial use.

**Reply by bkev13** [reply]
Thanks a lot for your help this gave me a great start for my project. I will mark the issue as answered for now and maybe (hopefully not) reopen it if I have more questions about this topic!

**Reply by bkev13** [reply]
Hi @sdhiscocks, I have another question: How can I get the content of the tracks set (for instance how many tracks got initiated, and deleted in total, what/how many measurements belong to each selected track)? It takes ages to print it, which I find odd given how fast it can plot the plots.
